relay/prover.go

```diff
@@ -166,25 +180,98 @@ func (pr *Prover) CreateInitialLightClientState(ctx context.Context, height expo
+			"height", header.GetHeight().String())
+
+		return header, nil
+	} else {
+		pr.getLogger().InfoContext(ctx, "using origin prover for latest finalized header")
+		return pr.originProver.GetLatestFinalizedHeader(ctx)
+	}
 }
 
 // SetupHeadersForUpdate returns the finalized header and any intermediate headers needed to apply it to the client on the counterparty chain
 // The order of the returned header slice should be as: [<intermediate headers>..., <update header>]
 // if the header slice's length == nil and err == nil, the relayer should skip the update-client
 func (pr *Prover) SetupHeadersForUpdate(ctx context.Context, dstChain core.FinalityAwareChain, latestFinalizedHeader core.Header) (<-chan *core.HeaderOrError, error) {
-	if err := pr.UpdateEKIIfNeeded(ctx, dstChain); err != nil {
+	var results []*shfu_storage.UpdateClientResult
+	var err error
+
+	// Use SHFU gRPC server if configured (environment variable or config), otherwise use local implementation
+	useGRPC, grpcAddress := pr.shouldUseSHFUGRPC()
+	if useGRPC {
+		if err := pr.UpdateEKIIfNeeded(ctx, dstChain); err != nil {
+			return nil, err
+		}
+		results, err = getUpdateClientFromGRPC(ctx, pr.getLogger(), grpcAddress, pr.originChain, dstChain, latestFinalizedHeader)
+	} else {
+		pr.getLogger().InfoContext(ctx, "using local SHFU implementation")
+		results, err = pr.setupHeadersForUpdate0(ctx, dstChain, latestFinalizedHeader)
+	}
+
+	if err != nil {
 		return nil, err
 	}
 
+	if len(results) == 0 {
+		return core.MakeHeaderStream(), nil
+	}
+
+	// Extract messages and signatures from results for existing aggregation logic
+	var messages [][]byte
+	var signatures [][]byte
+	for _, result := range results {
+		messages = append(messages, result.Message)
+		signatures = append(signatures, result.Signature)
+	}
+
+	var updates []core.Header
+	// NOTE: assume that the messages length and the signatures length are the same
+	if pr.config.MessageAggregation {
+		pr.getLogger().InfoContext(ctx, "aggregate messages", "num_messages", len(messages))
+		update, err := aggregateMessages(ctx, pr.getLogger(), pr.config.GetMessageAggregationBatchSize(), pr.lcpServiceClient.AggregateMessages, messages, signatures, pr.activeEnclaveKey.GetEnclaveKeyAddress().Bytes())
+		if err != nil {
+			return nil, err
+		}
+		updates = append(updates, update)
+	} else {
+		pr.getLogger().InfoContext(ctx, "updateClient", "num_messages", len(messages))
+		for i := 0; i < len(messages); i++ {
+			updates = append(updates, &lcptypes.UpdateClientMessage{
+				ProxyMessage: messages[i],
+				Signatures:   [][]byte{signatures[i]},
+			})
+		}
+	}
+	return core.MakeHeaderStream(updates...), nil
+}
+
+// setupHeadersForUpdate0 performs the initial setup and updateClient calls
+// Returns the processed updateClient results for aggregation
+func (pr *Prover) setupHeadersForUpdate0(ctx context.Context, dstChain core.FinalityAwareChain, latestFinalizedHeader core.Header) ([]*shfu_storage.UpdateClientResult, error) {
```

Comment:
I think `updateELCForUpdateClient` is more appropriate:

```suggestion
func (pr *Prover) updateELCForUpdateClient(ctx context.Context, dstChain core.FinalityAwareChain, latestFinalizedHeader core.Header) ([]*shfu_storage.UpdateClientResult, error) {
```

Reply:
renamed